BuildScript/BuildTasks/TestConfigurationsFilter.cs

```diff
@@ -0,0 +1,291 @@
+        IReadOnlyCollection<string> enabledPlatforms,
+        IReadOnlyCollection<string> supportedExecutionRuntimes,
+        IReadOnlyCollection<string> enabledExecutionRuntimes,
+        IReadOnlyCollection<string> supportedTargetRuntimes,
+        IReadOnlyCollection<string> enabledTargetRuntimes,
+        IReadOnlyCollection<string> supportedConfigurationIDs,
+        IReadOnlyCollection<string> enabledConfigurationIDs,
+        ITaskLogger logger)
+    {
+      _supportedBrowsers = supportedBrowsers;
+      _enabledBrowsers = enabledBrowsers;
+      _supportedDatabaseSystems = supportedDatabaseSystems;
+      _enabledDatabaseSystems = enabledDatabaseSystems;
+      _supportedPlatforms = supportedPlatforms;
+      _enabledPlatforms = enabledPlatforms;
+      _enabledExecutionRuntimes = enabledExecutionRuntimes;
+      _supportedTargetRuntimes = supportedTargetRuntimes;
+      _enabledTargetRuntimes = enabledTargetRuntimes;
+      _supportedConfigurationIDs = supportedConfigurationIDs;
+      _enabledConfigurationIDs = enabledConfigurationIDs;
+      _logger = logger;
+      _dockerImageToKeyDictionary = GetImageToKeyDictionary (supportedExecutionRuntimes);
+    }
+
+    public IReadOnlyCollection<TestConfiguration> GetFilteredTestConfigurations (IReadOnlyCollection<TestConfiguration> testConfigurationsToFilter)
+    {
+      EnsureNoEnabledMetadataIsEmpty();
+      EnsureAllMetadataAreSupported (testConfigurationsToFilter);
+
+      var filteredItems = GetFilteredItems (testConfigurationsToFilter).ToArray();
+
+      var distinctInputAssemblyPaths = GetTestAssemblyFullPaths (testConfigurationsToFilter).Distinct();
+
+      foreach (var testAssemblyPath in distinctInputAssemblyPaths)
+      {
+        if (!GetTestAssemblyFullPaths (filteredItems).Contains (testAssemblyPath))
+          _logger.LogWarning ($"All testing configurations for {testAssemblyPath} have been filtered out. No tests will be run for that assembly.");
+      }
+
+      return filteredItems;
+    }
+
+    private Dictionary<string, string> GetImageToKeyDictionary (IReadOnlyCollection<string> supportedExecutionRuntimes)
+    {
+      try
+      {
+        return supportedExecutionRuntimes
+            .Select (x => x.Split (new[] { "::" }, StringSplitOptions.RemoveEmptyEntries))
+            .ToDictionary (split => split[1], split => split[0], StringComparer.OrdinalIgnoreCase);
+      }
+      catch (ArgumentException ex) when (ex.Message == "An item with the same key has already been added.")
+      {
+        _logger.LogError ("It is not possible to assign a docker image to multiple keys.");
+        throw;
+      }
+    }
+
+    private void EnsureNoEnabledMetadataIsEmpty ()
+    {
+      var errorsForEmptyEnabledMetadata = GetErrorsForEmptyEnabledMetadata().ToArray();
+
+      foreach (var error in errorsForEmptyEnabledMetadata)
+        _logger.LogError (error);
+
+      if (errorsForEmptyEnabledMetadata.Any())
+        throw new InvalidOperationException (
```

Comment:
Honestly, this as somewhat of a design smell that came from turning two separate msbuild tasks into "normal" C# classes. They still need to do the logging, since they don't return whole collection of errors for the caller (e.g. when multiple properties are invalid, they log each one individually instead of logging once with all of them), so they log each error but also have to stop doing what they're doing once they know there are errors at all.

What I was actually thinking about was let the factory and the filter only throw exceptions (sometimes `AggregateException` for multiple errors) and let the msbuild task log the exception message. Do you think that would be better? I don't suppose it would take too long to write it.